File: 03-builder/03-builder.py
class Computer:
    def __init__(self, ram, cpu, ssd):
        self.ram = ram
        self.cpu = cpu
        self.ssd = ssd

    def stats(self):
        return f"ram is: {self.ram}GB, cpu is: {self.cpu}, ssd is: {self.ssd}."

class ComputerBuilder:
    def __init__(self):
        self.ram = None
        self.cpu = None
        self.ssd = None

    def set_ram(self, ram):
        self.ram = ram
        return self

    def set_cpu(self, cpu):
        self.cpu = cpu
        return self

    def set_ssd(self, ssd):
        self.ssd = ssd
        return self

    def build(self):
        return Computer(self.ram, self.cpu, self.ssd)

class Director:
    def __init__(self, builder: ComputerBuilder):
        self.builder = builder

    def home_pc(self):
        return self.builder.set_ssd("512").set_cpu("3.14").set_ram("8").build()
        
def main():
    builder = ComputerBuilder()
    director = Director(builder)
    home_pc = director.home_pc()
    print(home_pc.stats())

if __name__ == "__main__":
    main()

File: 03-builder/03-builder.js
class Computer  {
    constructor(ram, cpu, ssd) {
        this.ram = ram,
        this.cpu = cpu,
        this.ssd = ssd
    }

    stats() {
        console.log('ram: ', this.ram, 'cpu: ', this.cpu, 'ssd: ', this.ssd)
    }
}

class ComputerBuilder {
    constructor() {
        this.ram = ""
        this.cpu = ""
        this.ssd = ""
    }

    setRam(ram) {
        this.ram = ram;
        return this;
    }

    setCpu(cpu) {
        this.cpu = cpu;
        return this;
    }

    setSsd(ssd) {
        this.ssd = ssd;
        return this;
    }

    build() {
        return new Computer(this.ram, this.cpu, this.ssd);
    }
}

class Director {
    constructor(Builder) {
        this.builder = Builder
    }

    homePc() {
        return this.builder
            .setRam("24GB")
            .setCpu("5.1GHz")
            .setSsd("1TB")
            .build()
    }
}

const builder = new ComputerBuilder();
const director = new Director(builder);

homePc = director.homePc();
homePc.stats();
